packages/tpetra/core/test/FECrsMatrix/FECrsMatrix_UnitTests.cpp

```diff
@@ -301,15 +302,97 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL( FECrsMatrix, Assemble1D, LO, GO, Scalar, Node
   TPETRA_GLOBAL_SUCCESS_CHECK(out,comm,success)
 }
 
-////
-TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL( FECrsMatrix, Assemble1D_Kokkos, LO, GO, Scalar, Node )
-{
+//// specializations for kokkos fecrsmatrix assembly
+//std::complex<double>
+template<class LO, class GO, class Scalar, class Node>  
+bool assemble1DKokkos(Teuchos::FancyOStream &out, std::complex<double> t) {
+  using exec_space = typename Node::execution_space;
+  using range_type = Kokkos::RangePolicy<exec_space, LO>;
+  using FEMAT = typename Tpetra::FECrsMatrix<std::complex<double>,LO,GO,Node>;
+  using CMAT = typename Tpetra::CrsMatrix<std::complex<double>,LO,GO,Node>;
+  using FEG = typename Tpetra::FECrsGraph<LO,GO,Node>;
+
+  // get a comm
+  RCP<const Comm<int> > comm = getDefaultComm();
+  
+  // Generate a mesh
+  size_t numLocal = 10;
+  GraphPack<LO,GO,Node> pack;
+  generate_fem1d_graph(numLocal,comm,pack);
+
+  // Make the graph    
+  // FIXME: We should be able to get away with 3 for StaticProfile here, but we need 4 since duplicates are
+  // not being handled correctly. 
+  RCP<FEG> graph = rcp(new FEG(pack.uniqueMap,pack.overlapMap,4));
+
+  graph->beginFill();
+  for(size_t i=0; i<(size_t)pack.element2node.size(); i++) {
+    for(size_t j=0; j<pack.element2node[i].size(); j++) {
+      GO gid_j = pack.element2node[i][j];
+      for(size_t k=0; k<pack.element2node[i].size(); k++) {
+        GO gid_k = pack.element2node[i][k];
+        //printf("Inserting (%d,%d)\n",gid_j,gid_k);
+        graph->insertGlobalIndices(gid_j,1,&gid_k);
+      }
+    }
+  }
+  graph->endFill();
+
+  // Generate the "local stiffness matrix"
+  std::vector<std::vector<std::complex<double>> > localValues = 
+    generate_fem1d_element_values<std::complex<double>>();
+  auto kokkosValues = generate_fem1d_element_values_kokkos<std::complex<double>, Node>();
+
+  // Make the matrix two ways
+  FEMAT mat1(graph); // Here we use graph as a FECrsGraph
+  CMAT mat2(graph);  // Here we use graph as a CrsGraph in OWNED mode
+  mat1.beginFill();
+
+  auto k_e2n = pack.k_element2node;
+  auto localMat = mat1.getLocalMatrix();
+  auto localMap = pack.overlapMap->getLocalMap();
+  //get local map too
+
+  Kokkos::parallel_for("assemble_1d",
+		       range_type (0,k_e2n.extent(0)), 
+		       KOKKOS_LAMBDA(const size_t i) {
+    size_t extent = k_e2n.extent(1);
+    for(size_t j=0; j < extent; j++) {
+      LO lid_j = localMap.getLocalElement(k_e2n(i, j));
+      for(size_t k=0; k < extent; k++) {
+        LO lid_k = localMap.getLocalElement(k_e2n(i, k));
+	fprintf(stderr, "converting to kokkos complex\n");
+	Kokkos::complex<double> elem(kokkosValues(j, k));
```

Comment:
[**[complex.numbers]**](https://eel.is/c++draft/complex.numbers) in C++11 specifies that you may freely `reinterpret_cast` between `std::complex<T>` and `T[2]`.  We use this to justify freely `reinterpret_cast`ing between `Kokkos::complex<T>` (or `impl_scalar_type`) and `std::complex<T>` (or `scalar_type`).  This means that you can write generic code on Scalar, as long as you respect that `impl_scalar_type` and `scalar_type` may not necessarily be the same type.  Thus, you should not need to write a complex-arithmetic specialization of a test, unless the test exercises something specific about complex numbers.